include/libsemigroups/konieczny.hpp

```diff
@@ -0,0 +1,3522 @@
+    // not noexcept because operator++ isn't necessarily
+    const_d_class_iterator cbegin_D_classes() const {
+      auto it = _D_classes.cbegin();
+      return const_d_class_iterator(it)
+             + ((_D_classes.size() > 0 && _adjoined_identity_contained) ? 0
+                                                                        : 1);
+    }
+
+    //! Returns a const iterator referring to past the pointer to the last
+    //! \f$\mathscr{D}\f$-class of the semigroup.
+    //!
+    //! This member function does not perform any enumeration of the
+    //! semigroup; the iterator returned may be invalidated by any call to a
+    //! non-const member function of the Konieczny class.
+    const_d_class_iterator cend_D_classes() const noexcept {
+      return const_d_class_iterator(_D_classes.cend());
+    }
+
+    //! A type for const iterators through the regular \f$\mathscr{D}\f$-classes
+    //! of \c this, in the order they were enumerated.
+    //!
+    //! \sa const_d_class_iterator.
+    using const_regular_d_class_iterator
+        = detail::ConstIteratorStateless<DClassIteratorTraits<RegularDClass>>;
+
+    //! Returns a const iterator referring to a pointer to the first regular
+    //! D- class of the semigroup.
+    //!
+    //! This member function does not perform any enumeration of the
+    //! semigroup; the iterator returned may be invalidated by any call to a
+    //! non-const member function of the Konieczny class.
+    // not noexcept because operator++ isn't necessarily
+    const_regular_d_class_iterator cbegin_regular_D_classes() const {
+      auto it = _regular_D_classes.cbegin();
+      return const_regular_d_class_iterator(it)
+             + ((_regular_D_classes.size() > 0 && _adjoined_identity_contained)
+                    ? 0
+                    : 1);
+    }
+
+    const_regular_d_class_iterator cbegin_rdc() const noexcept {
+      return cbegin_regular_D_classes();
+    }
+
+    //! Returns a const iterator referring to past the pointer to the last
+    //! regular \f$\mathscr{D}\f$-class of the semigroup.
+    //!
+    //! This member function does not perform any enumeration of the
+    //! semigroup; the iterator returned may be invalidated by any call to a
+    //! non-const member function of the Konieczny class.
+    const_regular_d_class_iterator cend_regular_D_classes() const noexcept {
+      return const_regular_d_class_iterator(_regular_D_classes.cend());
+    }
+
+    const_regular_d_class_iterator cend_rdc() const {
+      return cend_regular_D_classes();
+    }
+
+   private:
+    using CacheGuard = detail::CacheGuard<internal_element_type>;
+
+    ////////////////////////////////////////////////////////////////////////
+    // Konieczny - utility methods - private
+    ////////////////////////////////////////////////////////////////////////
+
+    // TODO(later): can we do this differently to avoid being a friend?
+    std::vector<internal_element_type>& covering_reps(DClass& D) {
```

Comment:
I was trying to avoid this so that `Konieczny` wouldn't have to be a friend of `DClass` just for the sake of `covering_reps`, but I've had to make it a friend for other reasons anyway.